students/mkdir01/lesson05/mailroom3.py

```diff
@@ -0,0 +1,105 @@
+        elif name == "":
+            return
+        else:
+            dollars = input("Please enter a donation amount (ex: 500.05)> ")
+            for k in data:
+                if k == name:
+                    try:
+                        data[k].append(round(float(dollars), 2))
+                        break
+                    except ValueError:
+                        print("Donation amounts must be numbers")
+                        if not data[name]:
+                            data.pop(name)  # undo: data[name] = []
+            else:  # if no name match, make new entry
+                data[name] = [(round(float(dollars), 2))]
+            print(f"Dear {name},\n\nThank you for your donation of {data[name][-1]:.2f}.\n\nLove,\nThe Charity Org\n")
+            return
+
+
+# mode 2 - "create a report"
+def sort_by_dollars(donor_lists):
+    try:
+        donor_lists.sort(key=lambda k: k[1], reverse=True)
+    except IndexError:
+        print('This list can not be sorted.')
+        print(donor_lists)
+
+
+def create_report():
+    # create [[donor 1 name, donor 1 total donations], [donor 2 name, donor 2 total donations], etc]
+    temp_donor_lists = [[k, round(float(sum(data[k])), 2)] for k in data]
+    sort_by_dollars(temp_donor_lists)
+
+    # create [[donor name, total donations, # gifts, avg gift], [donor 2 info]]
+    for l in temp_donor_lists:
+        l.append(len(data[l[0]]))  # append number of gifts
+        try:
+            l.append(round(l[1] / (len(data[l[0]])), 2))  # append avg gift)
+        except ZeroDivisionError:
+            temp_donor_lists.pop(l)  # gets rid of any rows that have 0 donations
+
+    print("Donor Name                | Total Given |   Num Gifts |  Average Gift")
+    print("---------------------------------------------------------------------")
+    for row in temp_donor_lists:
+        print("{:<27}{:>14.2f}{:>14}{:>14.2f}".format(*row))
+
+
+# mode 3 - "send letters"
+def plural_donate(n):
+    if n > 1:
+        return 'donations totaling'
+    else:
+        return 'donatation of'
+
+
+def total_donate(n):
+    total = 0
+    for d in n:
+        total += d
+    return round(float(total), 2)
+
+
+def send_letters():
+    for k in data:
+        with open(k + '.txt', 'w') as outfile:
+            outfile.write(f"Dear {k},\n\tThank you for your {plural_donate(len(data[k]))} {total_donate(data[k]):.2f}.\n\tIt will be put to good use.\n\n\tSincerely,\n\t- The Team")
+
+
+# mode 4 - "quit"
+def end_program():  # exit() and quit() look like they're reserved
+    print("Thanks for playing along.")
+    sys.exit()
+
+
+# main loop
+def main():
+    menu_dict = {"1": thank_you, "2": create_report, "3": send_letters, "4": end_program}
+    while True:
+        print("Choose a number:")
+        print("(1) Send a Thank You")
+        print("(2) Create a Report")
+        print("(3) Send letter to everyone")
+        print("(4) quit")
+        mode = input("> ")
+        if mode in menu_dict:
```

Comment:
If you use bracket syntax (`menu_dict[mode]`) to access the dispatch dict, you could use a try/except for `KeyError` here to catch bad user input